File: swift/100_SameTree.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func isSameTree(_ p: TreeNode?, _ q: TreeNode?) -> Bool {
        var pNodes: [TreeNode?] = [p], qNodes: [TreeNode?] = [q]
        
        while !pNodes.isEmpty && !qNodes.isEmpty {
            var p = pNodes.removeLast()
            var q = qNodes.removeLast()
            
            if p == nil && q == nil { continue }
            else if p == nil || q == nil || (p!.val != q!.val) { return false }
            
            pNodes.append(p?.left)
            pNodes.append(p?.right)
            
            qNodes.append(q?.left)
            qNodes.append(q?.right)
        }
        
        return true
    }
}

File: python/100_SameTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        if not (p and q):
            return True if p == q else False
        if p.val != q.val:
            return False
        
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
        

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    isTrue = True
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        if not p and not q:
            return True
        elif not p or not q:
            return False
        
        if p.val != q.val:
            return False
        
        if p.left or q.left and self.isTrue:
            self.isTrue = self.isSameTree(p.left, q.left)
        
        if p.right or q.right and self.isTrue:
            self.isTrue = self.isSameTree(p.right, q.right)
            
        return self.isTrue
